### PR TITLE
[SPARK-25248] [CORE] Audit barrier Scala APIs for 2.4

### PR DESCRIPTION
**Pull request by mengxr**
## What changes were proposed in this pull request?

I made one pass over barrier APIs added to Spark 2.4 and updates some scopes and docs. I will update Python docs once Scala doc was reviewed.

One major issue is that `BarrierTaskContext` implements `TaskContextImpl` that exposes some public methods. And internally there were several direct references to `TaskContextImpl` methods instead of `TaskContext`. This PR moved some methods from `TaskContextImpl` to `TaskContext`, remaining package private, and used delegate methods to avoid inheriting `TaskContextImp` and exposing unnecessary APIs.

TODOs:
- [x] scala doc
- [x] python doc (#22261 ).

### CODE REVIEW

**Comment by jiangxb1987** [comment]
retest this please

**Review by jiangxb1987 — APPROVED** [review]
LGTM

File: core/src/main/scala/org/apache/spark/BarrierTaskContext.scala [hunk 1]
```diff
@@ -68,7 +74,7 @@ class BarrierTaskContext(
    *
    * CAUTION! In a barrier stage, each task must have the same number of barrier() calls, in all
    * possible code branches. Otherwise, you may get the job hanging or a SparkException after
-   * timeout. Some examples of misuses listed below:
+   * timeout. Some examples of '''misuses''' listed below:
```
[the remark this answers is not in the document]
Comment (kiszk) [reply core/src/main/scala/org/apache/spark/BarrierTaskContext.scala]: nit: `listed` -> `are listed`?

**Comment by mengxr** [comment]
test this please

File: core/src/main/scala/org/apache/spark/BarrierTaskContext.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (mengxr) [reply core/src/main/scala/org/apache/spark/BarrierTaskContext.scala]: just saw it, will include it if Jenkins fails:)

**Comment by kiszk** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/scheduler/Task.scala [hunk 2]
```diff
@@ -82,31 +82,22 @@ private[spark] abstract class Task[T](
     SparkEnv.get.blockManager.registerTask(taskAttemptId)
     // TODO SPARK-24874 Allow create BarrierTaskContext based on partitions, instead of whether
     // the stage is barrier.
-    context = if (isBarrier) {
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: `context` is still used in the following statements.

**Comment by jiangxb1987** [comment]
LGTM

**Comment by mengxr** [comment]
Merged into master. Thanks for review!

**Pull request closed without merging**